README.md
# [gatsby-plugin-google-gtag-optin](https://www.npmjs.com/package/gatsby-plugin-google-gtag-optin)

This is a fork of the official Gatsby plugin [gatsby-plugin-google-gtag](https://www.gatsbyjs.org/packages/gatsby-plugin-google-gtag/). It's supposed to be used as a drop in replacement.

It functions the same but **ONLY** after the user has consented. You can signal consent by setting a localStorage optinKey (default is `gtag_optin`) to `true` and calling `window.loadGtag()` to immediately load the Google Global Site Tag. If you set the key in localStorage you don't have to call `window.loadGtag()` on subsequent visits.

**Example:**

```js
import { GTAG_OPTIN_KEY } from "gatsby-plugin-google-gtag-optin" // Or use your own if you changed it in the config

const consent = () => {
  localStorage.setItem(GTAG_OPTIN_KEY, true)
  if (typeof window.loadGtag == "function") {
    window.loadGtag()
  }
}
```

---

Easily add Google Global Site Tag to your Gatsby site.

> The global site tag (gtag.js) is a JavaScript tagging framework and API that allows you to send event data to Google Analytics, Google Ads, Campaign Manager, Display & Video 360, and Search Ads 360.

Global site tag (gtag.js) is meant to combine multiple Google tagging systems and can replace older ones such as [analytics.js](https://developers.google.com/analytics/devguides/collection/analyticsjs/) ([gatsby-plugin-google-analytics](https://www.gatsbyjs.org/packages/gatsby-plugin-google-analytics/)).

For more general information on gtag you can read Google's official documentation on the subject: <https://developers.google.com/gtagjs/.>

If you're migrating from analytics.js (gatsby-plugin-google-analytics) you can read about the subtle API differences in more depth at: <https://developers.google.com/analytics/devguides/collection/gtagjs/migration.>

NOTE: This plugin only works in production mode! To test your Global Site Tag is installed and firing events correctly run: `gatsby build && gatsby serve.`

## Install

`npm install gatsby-plugin-google-gtag`

## How to use

```js
// In your gatsby-config.js
module.exports = {
  plugins: [
    {
      resolve: `gatsby-plugin-google-gtag`,
      options: {
        // You can add multiple tracking ids and a pageview event will be fired for all of them.
        trackingIds: [
          "GA-TRACKING_ID", // Google Analytics / GA
          "AW-CONVERSION_ID", // Google Ads / Adwords / AW
          "DC-FLOODIGHT_ID", // Marketing Platform advertising products (Display & Video 360, Search Ads 360, and Campaign Manager)
        ],
        // This object gets passed directly to the gtag config command
        // This config will be shared across all trackingIds
        gtagConfig: {
          optimize_id: "OPT_CONTAINER_ID",
          anonymize_ip: true,
          cookie_expires: 0,
        },
        // This object is used for configuration specific to this plugin
        pluginConfig: {
          // Puts tracking script in the head instead of the body
          head: false,
          // Setting this parameter is also optional
          respectDNT: true,
          // Avoids sending pageview hits from custom paths
          exclude: ["/preview/**", "/do-not-track/me/too/"],
          // localStorage key that has to be set to true for plugin to load
          optinKey: "gtag_optin", // default
        },
      },
    },
  ],
}
```

## Custom Events

This plugin automatically sends a "pageview" event to all products given as "trackingIds" on every Gatsbys route change.

If you want to call a custom event you have access to `window.gtag` where you can call an event for all products:

```js
window.gtag("event", "click", { ...data })
```

or you can target a specific product:

```js
window.gtag("event", "click", { send_to: "AW-CONVERSION_ID", ...data })
```

In either case don't forget to guard against SSR:

```js
typeof window !== "undefined" && window.gtag("event", "click", { ...data })
```

## `<OutboundLink>` component

To make it easy to track clicks on outbound links the plugin provides a component.

To use it, simply import it and use it like you would the `<a>` element e.g.

```jsx
import React from "react"
import { OutboundLink } from "gatsby-plugin-google-gtag"

export default () => (
  <div>
    <OutboundLink href="https://www.gatsbyjs.org/packages/gatsby-plugin-google-gtag/">
      Visit the Google Global Site Tag plugin page!
    </OutboundLink>
  </div>
)
```

## The "gtagConfig.anonymize_ip" option

Some countries (such as Germany) require you to use the
[\_anonymizeIP](https://support.google.com/analytics/answer/2763052) function for
Google Site Tag. Otherwise you are not allowed to use it. The option adds the
block of code below:

```js
function gaOptout() {
  ;(document.cookie =
    disableStr + "=true; expires=Thu, 31 Dec 2099 23:59:59 UTC;path=/"),
    (window[disableStr] = !0)
}

var gaProperty = "UA-XXXXXXXX-X",
  disableStr = "ga-disable-" + gaProperty
document.cookie.indexOf(disableStr + "=true") > -1 && (window[disableStr] = !0)
```

If your visitors should be able to set an Opt-Out-Cookie (No future tracking)
you can set a link e.g. in your imprint as follows:

`<a href="javascript:gaOptout();">Deactivate Google Tracking</a>`

## The "gtagConfig.optimize_id" option

If you need to use Google Optimize for A/B testing, you can add this optional Optimize container id to allow Google Optimize to load the correct test parameters for your site.

## Other "gtagConfig" options

The `gtagConfig` is passed directly to the gtag config command, so you can specify everything it supports, e.g. `gtagConfig.cookie_name`, `gtagConfig.sample_rate`. If you're migrating from the analytics.js plugin, this means that all Create Only Fields should be snake_cased.

## The "pluginConfig.respectDNT" option

If you enable this optional option, Google Global Site Tag will not be loaded at all for visitors that have "Do Not Track" enabled. While using Google Global Site Tag does not necessarily constitute Tracking, you might still want to do this to cater to more privacy oriented users.

## The "pluginConfig.exclude" option

If you need to exclude any path from the tracking system, you can add it (one or more) to this optional array as glob expressions.


CHANGELOG.md
# Changelog: `gatsby-plugin-google-gtag`

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [4.0.0](https://github.com/gatsbyjs/gatsby/commits/HEAD/packages/gatsby-plugin-google-gtag) (2021-10-21)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v4.0)

**Note:** Version bump only for package gatsby-plugin-google-gtag

## [3.14.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.14.0/packages/gatsby-plugin-google-gtag) (2021-09-18)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.14)

#### Chores

- update babel monorepo [#32996](https://github.com/gatsbyjs/gatsby/issues/32996) ([048c7a7](https://github.com/gatsbyjs/gatsby/commit/048c7a727bbc6a9ad8e27afba72ee20e946c4aaa))

## [3.13.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.13.0/packages/gatsby-plugin-google-gtag) (2021-09-01)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.13)

#### Chores

- re-generate changelogs [#32886](https://github.com/gatsbyjs/gatsby/issues/32886) ([417df15](https://github.com/gatsbyjs/gatsby/commit/417df15230be368a9db91f2ad1a9bc0442733177))

## [3.12.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.12.0/packages/gatsby-plugin-google-gtag) (2021-08-18)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.12)

#### Chores

- update babel monorepo [#32564](https://github.com/gatsbyjs/gatsby/issues/32564) ([a554998](https://github.com/gatsbyjs/gatsby/commit/a554998b4f6765103b650813cf52dbfcc575fecf))

## [3.11.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.11.0/packages/gatsby-plugin-google-gtag) (2021-08-04)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.11)

**Note:** Version bump only for package gatsby-plugin-google-gtag

## [3.10.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.10.0/packages/gatsby-plugin-google-gtag) (2021-07-20)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.10)

#### Bug Fixes

- replace google analytics domain with google tag manager [#31036](https://github.com/gatsbyjs/gatsby/issues/31036) ([5afbda0](https://github.com/gatsbyjs/gatsby/commit/5afbda0922da992ff2446cba4c18a772df65e3c6))

#### Chores

- update babel monorepo [#32238](https://github.com/gatsbyjs/gatsby/issues/32238) ([466d4c0](https://github.com/gatsbyjs/gatsby/commit/466d4c087bbc96abb942a02c67243bcc9a4f2a0a))

## [3.9.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.9.0/packages/gatsby-plugin-google-gtag) (2021-07-07)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.9)

**Note:** Version bump only for package gatsby-plugin-google-gtag

## [3.8.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.8.0/packages/gatsby-plugin-google-gtag) (2021-06-23)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.8)

#### Chores

- bump babel minor [#31857](https://github.com/gatsbyjs/gatsby/issues/31857) ([7d42e8d](https://github.com/gatsbyjs/gatsby/commit/7d42e8d866e46e9c39838d812d080d06433f7060))

### [3.7.1](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.7.1/packages/gatsby-plugin-google-gtag) (2021-06-10)

#### Chores

- bump babel minor [#31857](https://github.com/gatsbyjs/gatsby/issues/31857) [#31859](https://github.com/gatsbyjs/gatsby/issues/31859) ([8636025](https://github.com/gatsbyjs/gatsby/commit/863602567930a39142ed33d9d1f1813b7dec8686))

## [3.7.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.7.0/packages/gatsby-plugin-google-gtag) (2021-06-09)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.7)

#### Chores

- update babel monorepo [#31143](https://github.com/gatsbyjs/gatsby/issues/31143) ([701ab2f](https://github.com/gatsbyjs/gatsby/commit/701ab2f6690c3f1bbaf60cf572513ea566cc9ec9))

## [3.6.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.6.0/packages/gatsby-plugin-google-gtag) (2021-05-25)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.6)

#### Chores

- Fix typo Fix [#31431](https://github.com/gatsbyjs/gatsby/issues/31431) ([be8fc87](https://github.com/gatsbyjs/gatsby/commit/be8fc8738ce7c704348cd6c8c79e2a86e74d8455))

## [3.5.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.5.0/packages/gatsby-plugin-google-gtag) (2021-05-12)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.5)

**Note:** Version bump only for package gatsby-plugin-google-gtag

## [3.4.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.4.0/packages/gatsby-plugin-google-gtag) (2021-04-28)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.4)

**Note:** Version bump only for package gatsby-plugin-google-gtag

## [3.3.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.3.0/packages/gatsby-plugin-google-gtag) (2021-04-14)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.3)

**Note:** Version bump only for package gatsby-plugin-google-gtag

## [3.2.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.2.0/packages/gatsby-plugin-google-gtag) (2021-03-30)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.2)

**Note:** Version bump only for package gatsby-plugin-google-gtag

## [3.1.0](https://github.com/gatsbyjs/gatsby/commits/gatsby-plugin-google-gtag@3.1.0/packages/gatsby-plugin-google-gtag) (2021-03-16)

[🧾 Release notes](https://www.gatsbyjs.com/docs/reference/release-notes/v3.1)

**Note:** Version bump only for package gatsby-plugin-google-gtag


src/__tests__/gatsby-ssr.js
import React from "react"
import { onRenderBody } from "../gatsby-ssr"

it(`does not crash when no pluginConfig is provided`, () => {
  const mocks = {
    setHeadComponents: () => null,
    setPostBodyComponents: () => null,
  }

  const pluginOptions = {
    trackingIds: [`GA_TRACKING_ID`],
  }
  onRenderBody(mocks, pluginOptions)
})

const DO_NOT_TRACK_STRING = `!(navigator.doNotTrack == "1" || window.doNotTrack == "1")`

it(`adds a preconnect link for Google Tag Manager`, () => {
  const mocks = {
    setHeadComponents: jest.fn(),
    setPostBodyComponents: jest.fn(),
  }
  const pluginOptions = {
    trackingIds: [`GA_TRACKING_ID`],
    pluginConfig: {},
  }

  onRenderBody(mocks, pluginOptions)
  const [link] = mocks.setHeadComponents.mock.calls[0][0]

  expect(link).toEqual(
    <link
      rel="preconnect"
      key="preconnect-google-gtag"
      href="https://www.googletagmanager.com"
    />,
    <link
      rel="dns-prefetch"
      key="dns-prefetch-google-gtag"
      href="https://www.googletagmanager.com"
    />
  )
})

describe(`respectDNT`, () => {
  it(`defaults respectDNT option to false`, () => {
    const mocks = {
      setHeadComponents: jest.fn(),
      setPostBodyComponents: jest.fn(),
    }
    const pluginOptions = {
      trackingIds: [`GA_TRACKING_ID`],
      pluginConfig: {},
    }

    onRenderBody(mocks, pluginOptions)
    const [, config] = mocks.setPostBodyComponents.mock.calls[0][0]

    expect(config.props.dangerouslySetInnerHTML.__html).not.toContain(
      DO_NOT_TRACK_STRING
    )
  })

  it(`listens to respectDNT option`, () => {
    const mocks = {
      setHeadComponents: jest.fn(),
      setPostBodyComponents: jest.fn(),
    }
    const pluginOptions = {
      trackingIds: [`GA_TRACKING_ID`],
      pluginConfig: {
        respectDNT: true,
      },
    }

    onRenderBody(mocks, pluginOptions)
    const [, config] = mocks.setPostBodyComponents.mock.calls[0][0]

    expect(config.props.dangerouslySetInnerHTML.__html).toContain(
      DO_NOT_TRACK_STRING
    )
  })
})


.eslintrc.js
module.exports = {
  env: {
    browser: true,
    es6: true,
    node: true,
    jest: true,
  },
  extends: [`eslint:recommended`, `plugin:react/recommended`],
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: `module`,
    ecmaFeatures: {
      jsx: true,
    },
  },
  plugins: [`react`, `prettier`],
  rules: {
    "no-invalid-this": `off`,
    "arrow-body-style": [
      `error`,
      `as-needed`,
      { requireReturnForObjectLiteral: true },
    ],
    "new-cap": `off`,
    "no-unused-vars": [
      `warn`,
      {
        varsIgnorePattern: `^_`,
        argsIgnorePattern: `^_`,
        ignoreRestSiblings: true,
      },
    ],
    "consistent-return": [`error`],
    "no-console": `off`,
    "no-inner-declarations": `off`,
    "prettier/prettier": `error`,
    quotes: [`error`, `backtick`],
    "react/display-name": `off`,
    "react/jsx-key": `warn`,
    "react/no-unescaped-entities": `off`,
    "react/prop-types": `off`,
    "require-jsdoc": `off`,
    "valid-jsdoc": `off`,
    "prefer-promise-reject-errors": `warn`,
    "no-prototype-builtins": `warn`,
    "guard-for-in": `warn`,
    "spaced-comment": [
      `error`,
      `always`,
      { markers: [`/`], exceptions: [`*`, `+`] },
    ],
    camelcase: [
      `error`,
      {
        properties: `never`,
        ignoreDestructuring: true,
        allow: [`^unstable_`],
      },
    ],
  },
  settings: {
    react: {
      version: `16.9.0`,
    },
  },
}


src/index.js
import React from "react"
import PropTypes from "prop-types"

const GTAG_OPTIN_KEY = `gtag_optin`

const OutboundLink = React.forwardRef(({ children, ...props }, ref) => (
  <a
    ref={ref}
    {...props}
    onClick={e => {
      if (typeof props.onClick === `function`) {
        props.onClick(e)
      }
      let redirect = true
      if (
        e.button !== 0 ||
        e.altKey ||
        e.ctrlKey ||
        e.metaKey ||
        e.shiftKey ||
        e.defaultPrevented
      ) {
        redirect = false
      }
      if (props.target && props.target.toLowerCase() !== `_self`) {
        redirect = false
      }
      if (window.gtag) {
        window.gtag(`event`, `click`, {
          event_category: `outbound`,
          event_label: props.href,
          transport_type: redirect ? `beacon` : ``,
          event_callback: function () {
            if (redirect) {
              document.location = props.href
            }
          },
        })
      } else {
        if (redirect) {
          document.location = props.href
        }
      }

      return false
    }}
  >
    {children}
  </a>
))

OutboundLink.propTypes = {
  href: PropTypes.string,
  target: PropTypes.string,
  onClick: PropTypes.func,
}

export { OutboundLink, GTAG_OPTIN_KEY }


src/gatsby-ssr.js
import React from "react"
import { Minimatch } from "minimatch"
import { GTAG_OPTIN_KEY } from "./index"

exports.onRenderBody = (
  { setHeadComponents, setPostBodyComponents },
  pluginOptions
) => {
  if (process.env.NODE_ENV !== `production` && process.env.NODE_ENV !== `test`)
    return null

  // Lighthouse recommends pre-connecting to google tag manager
  setHeadComponents([
    <link
      rel="preconnect"
      key="preconnect-google-gtag"
      href="https://www.googletagmanager.com"
    />,
    <link
      rel="dns-prefetch"
      key="dns-prefetch-google-gtag"
      href="https://www.googletagmanager.com"
    />,
  ])

  const gtagConfig = pluginOptions.gtagConfig || {}
  const pluginConfig = pluginOptions.pluginConfig || {}

  const OPTIN_KEY = pluginConfig.optinKey || GTAG_OPTIN_KEY

  // Prevent duplicate or excluded pageview events being emitted on initial load of page by the `config` command
  // https://developers.google.com/analytics/devguides/collection/gtagjs/#disable_pageview_tracking

  gtagConfig.send_page_view = false

  const firstTrackingId =
    pluginOptions.trackingIds && pluginOptions.trackingIds.length
      ? pluginOptions.trackingIds[0]
      : ``

  const excludeGtagPaths = []
  if (typeof pluginConfig.exclude !== `undefined`) {
    pluginConfig.exclude.map(exclude => {
      const mm = new Minimatch(exclude)
      excludeGtagPaths.push(mm.makeRe())
    })
  }

  const setComponents = pluginConfig.head
    ? setHeadComponents
    : setPostBodyComponents

  const renderHtml = () => `
      ${
        excludeGtagPaths.length
          ? `window.excludeGtagPaths=[${excludeGtagPaths.join(`,`)}];`
          : ``
      }
      ${
        typeof gtagConfig.anonymize_ip !== `undefined` &&
        gtagConfig.anonymize_ip === true
          ? `function gaOptout(){document.cookie=disableStr+'=true; expires=Thu, 31 Dec 2099 23:59:59 UTC;path=/',window[disableStr]=!0}var gaProperty='${firstTrackingId}',disableStr='ga-disable-'+gaProperty;document.cookie.indexOf(disableStr+'=true')>-1&&(window[disableStr]=!0);`
          : ``
      }
      if(${
        pluginConfig.respectDNT
          ? `!(navigator.doNotTrack == "1" || window.doNotTrack == "1")`
          : `true`
      }) {
        window.dataLayer = window.dataLayer || [];
        function gtag(){window.dataLayer && window.dataLayer.push(arguments);}
        gtag('js', new Date());
        ${pluginOptions.trackingIds
          .map(
            trackingId =>
              `gtag('config', '${trackingId}', ${JSON.stringify(gtagConfig)});`
          )
          .join(``)}
      }
      `

  const loadFunction = () => `
      function loadGtag() {
        const gtagScript = document.createElement("script")
        gtagScript.type = 'text/javascript'
        gtagScript.async = true
        gtagScript.src = "https://www.googletagmanager.com/gtag/js?id=${firstTrackingId}"
        const tag = document.getElementsByTagName('script')[0];
        tag.parentNode.insertBefore(gtagScript, tag)
      }
      if (localStorage.getItem("${OPTIN_KEY}") === "true") window.loadGtag()`

  return setComponents([
    <script
      key={`gatsby-plugin-google-gtag`}
      dangerouslySetInnerHTML={{ __html: loadFunction() }}
    />,
    <script
      key={`gatsby-plugin-google-gtag-config`}
      dangerouslySetInnerHTML={{ __html: renderHtml() }}
    />,
  ])
}
